File: python/1496.path-crossing.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        cur = (0, 0)
        cache = {cur}

        for move in path:
            match move:
                case 'N': cur = (cur[0], cur[1] + 1)
                case 'S': cur = (cur[0], cur[1] - 1)
                case 'E': cur = (cur[0] + 1, cur[1])
                case 'W': cur = (cur[0] - 1, cur[1])

            if cur in cache:
                return True

            cache.add(cur)

        return False

File: cpp/1496.path-crossing.cpp
class Solution {
public:
    bool isPathCrossing(string path) {
        set<pair<int, int>> cache;
        pair<int, int> cur = {0, 0};
        cache.insert(cur);

        for (auto dir: path) {
            switch (dir) {
                case 'N': cur.second++; break;
                case 'S': cur.second--; break;
                case 'E': cur.first++; break;
                case 'W': cur.first--; break;
            }
            if (cache.find(cur) != cache.end()) {
                return true;
            }
            cache.insert(cur);
        }
        return false;
    }
};